_posts/2021-02-04-do-not-trust-the-user-mental-model.md

```diff
@@ -0,0 +1,170 @@
+
+- The Customers Relationships owns the Customer concept
+- the Sales service owns the Product concept.
+
+> Don't hang up too much on names. They are not the point of the article.
+
+Assuming the above is correct, we immediately have a problem. To satisfy the new requirement, a hypothetical page showing products to our customers needs to query the customers owning service to gather the "IsPremium" attribute to determine the discount to apply to the displayed products' prices. The same applies if we move the discount calculation in the shopping cart or during the checkout phase.
+
+Regardless of where the discount calculation happens, coupling is served.
+
+## Reboot
+
+First, let's remove the names we assigned to the entities and start using fantasy names, such as colors. And instead of calling them entities, let's use the word service since the intention is to identify a set of attributes that makes sense together in the context of an owner, the service.
+
+```
+service Yellow
+{
+   string CustomerName;
+   string CustomerAddress;
+   bool IsCustomerPremium;
+}
+
+service Green
+{
+   string ProductName;
+   string ProductDescription;
+   decimal ProductPrice;
+}
+``` 
+
+At this point, what we can do is regroup attributes based on transactional boundaries. Another overloaded usage of the word transaction, sorry about that :-). The use case mentioned above and all the others we have in a real-world project are our guiding stars. Similarly to query patterns to satisfy when modeling No-SQL databases data structures, we can use behaviors described by requirements to group together attributes that interact and change together. They belong to the same transactional boundary.
+
+In the model that we're describing, based on the behaviors we need to implement, it's clear that there is no relationship between the customer name and the fact that it is a premium customer or not. The same applies to products; the product description doesn't influence the product price, and neither the other way around is correct. Although the product price indeed depends on the fact that a customer is a premium customer. Based on this observation, we can rearrange our clusters of attributes in the following way:
+
+```
+service Yellow
+{
+   string CustomerName;
+   string CustomerAddress;
+}
+
+service Green
+{
+   string ProductName;
+   string ProductDescription;
+}
+
+service Red
+{
+   bool IsCustomerPremium;
+   decimal ProductPrice;
+}
+```
+
+For the sake of the discussion, let's reapply names based on the user mental model to the above clusters. Yellow is still a customer, and Green is always a product. It's hard for Red to find a meaningful name; it could be something like "product price based on customer status." It's not a right entity name or a good table name; this is because it represents the set of data required by the described behavior, which changes over time and depends on the context.
```

Comment:
```suggestion
For the sake of the discussion, let's reapply names based on the user mental model to the above clusters. Yellow is still a customer, and Green is always a product. It's hard for Red to find a meaningful name; it could be something like "product price based on customer status." It's not a very good entity or table name; this is because it represents the set of data required by the described behavior, which changes over time and depends on the context.
```